### PR TITLE
Parse configuration file II: using serde, toml, merge & envy instead of config

### PR DESCRIPTION
**Pull request by robinkrahl**
In pull request #101, I used the config crate to parse and merge the configuration from files and from the environment.  config is apparently no longer maintained and uses an old version of the nom crate.  This PR replaces config with serde and toml for the configuration file parsing, merge for the merging of different configuration sources, and envy for reading the configuration from the environment.  Thus we no longer need the nom dependency and have more control over the dependency versions.

Note that we still have a separate `Config` struct that is manually merged with the `Args` struct.  We could also use the `Args` struct directly, but the question is how to handle the subcommand enums.  We could just skip them, or we could read the configuration for all subcommands into a map and then select the variant based on the CLI arguments.

In the first version of the patch, I used the app_dirs crate to locate the config directories, but it was unmaintained and had some annoying bugs and limitations (https://github.com/AndyBarron/app-dirs-rs/issues/34, https://github.com/AndyBarron/app-dirs-rs/issues/12).  We switched to the directories crate, but this crate does not support system configuration files.  Now there is a maintained fork of app_dirs, app_dirs2, and I prepared patches for the two app_dirs issues (https://github.com/app-dirs-rs/app_dirs2/pull/8, https://github.com/app-dirs-rs/app_dirs2/pull/12).  Once they are merged and released, I would move back to app_dirs2.

### CODE REVIEW

**Comment by d-e-s-o** [comment]
Thanks for the pull request Robin! I'll take a look this week.

File: src/config.rs [hunk 1]
```diff
@@ -25,21 +25,43 @@ use crate::Result;
...
 impl Config {
   pub fn load() -> Result<Self> {
-    load_user_config().map(|o| o.unwrap_or_else(Default::default))
+    use merge::Merge;
+
+    let mut config = Config::default();
+    if let Some(mut user_config) = load_user_config()? {
+        config.merge(&mut user_config);
+    }
+    config.merge(&mut envy::prefixed("NITROCLI_").from_env()?);
```
Comment (d-e-s-o) [inline src/config.rs:69]: Can you clarify how `envy` will behave if other variables with the `NITROCLI_` prefix are present but not used by the struct? Am I correct to assume that it will just ignore them?

File: doc/nitrocli.1 [hunk 2]
```diff
@@ -292,13 +330,27 @@ for the \fBuser\fR type.
...
+Restrict connections to the given device model (string, default none, see
+\fB\-\-model\fR).
 .TP
 .B NITROCLI_NO_CACHE
-If this variable is present in the environment, do not cache any inquired
-secrets using \fBgpg\-agent\fR(1) but ask for them each time they are needed.
-Note that this variable does not cause any cached secrets to be cleared. If a
-secret is already in the cache it will be ignored, but left otherwise untouched.
-Use the \fBpin clear\fR command to clear secrets from the cache.
+If set to true, do not cache any inquired secrets (boolean, default:
```
Comment (d-e-s-o) [inline doc/nitrocli.1:342]: Have you checked whether `true` is really the only option? Would `1` cause us to error out?

Can/should we keep the verbiage about not clearing existing cached secrets? That is, why did you remove so much?

**Comment by robinkrahl** [comment]
Thanks for the review!  Maybe I should have made it clearer that this is just a draft with many rough edges.  But I first wanted to reach consensus on the general approach before I polish the code.

File: CHANGELOG.md [hunk 3]
```diff
@@ -20,7 +26,6 @@
   root
 - Bumped `nitrokey` dependency to `0.6.0`
 - Bumped `quote` dependency to `1.0.3`
-- Bumped `syn` dependency to `1.0.14`
```
[the remark this answers is not in the document]
Comment (robinkrahl) [reply CHANGELOG.md:23]: Good catch, this is a rebase error.  (I wrote the original patches before v0.3.3 was released.)

File: Cargo.toml [hunk 4]
```diff
@@ -55,6 +55,9 @@ version = "0.4.1"
 [dependencies.libc]
 version = "0.2"
 
+[dependencies.merge]
+git = "https://git.sr.ht/~ireas/merge-rs"
```
[the remark this answers is not in the document]
Comment (robinkrahl) [reply Cargo.toml:62]: Of course!  AFAIK you can’t publish a crate on crates.io with Git dependencies.  But there are still some minor decisions to make that I’d like to discuss with you before I release something on crates.io.

File: src/args.rs [hunk 5]
```diff
@@ -57,6 +59,40 @@ impl From<DeviceModel> for nitrokey::Model {
   }
 }
 
+impl<'de> serde::Deserialize<'de> for DeviceModel {
```
[the remark this answers is not in the document]
Comment (robinkrahl) [reply src/args.rs:65]: Exactly.  If we move forward with this approach, I’m going to fix the macro so that we don’t need this implementation.

File: src/config.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
+// config.rs
+
+// *************************************************************************
+// * Copyright (C) 2020 Daniel Mueller (deso@posteo.net)                   *
```
[the remark this answers is not in the document]
Comment (robinkrahl) [reply src/config.rs:4]: Sounds good.  If you are changing this, would you also consider adding SPDX license annotations (see the [SPDX guide](https://spdx.github.io/spdx-spec/appendix-V-using-SPDX-short-identifiers-in-source-files/) and the [REUSE specification](https://reuse.software/spec/))?

File: src/config.rs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  /// The log level.
+  pub verbosity: u8,
+}
+
+impl Config {
+  pub fn load() -> Result<Self> {
+    load_user_config().map(|o| o.unwrap_or_else(Default::default))
+  }
+
+  pub fn update(&mut self, args: &args::Args) {
```
[the remark this answers is not in the document]
Comment (robinkrahl) [reply src/config.rs:74]: Exactly.  My `Merge` implementation only merges data structures of the same type.  We could just re-use the `Args` struct for the configuration, but that has other drawbacks, see the PR text.

File: src/config.rs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      self.model = args.model;
+    }
+    if args.verbose > 0 {
+      self.verbosity = args.verbose;
+    }
+  }
+}
+
+fn load_user_config() -> Result<Option<Config>> {
+    let path = path::PathBuf::from("config.toml");
```
[the remark this answers is not in the document]
Comment (robinkrahl) [reply src/config.rs]: Yes.  (This is only a temporary and removed in a later patch.)

[the remark this answers is not in the document]
Comment (robinkrahl) [reply src/config.rs]: Yes, the lookup mechanism is added [here](https://github.com/d-e-s-o/nitrocli/pull/111/commits/4c0a44269cb9bccefbfac54386b6735c5ba07cd9).

File: src/config.rs (hunk 1 above)
Comment (robinkrahl) [reply src/config.rs:69]: Yes.

File: doc/nitrocli.1 (hunk 2 above)
Comment (robinkrahl) [reply doc/nitrocli.1:342]: Yes, 1 causes an error: `provided string was not true or false while parsing value '1' provided by NO_CACHE`.

As I introduced a `--no-cache` option, I moved the detailed explanation to the documentation for that option.  The documentation for the environment variable refers to the option for more information.

File: src/config.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/config.rs:4]: Sure

File: src/args.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/args.rs:65]: Sounds good

**Comment by d-e-s-o** [comment]
> Thanks for the review! Maybe I should have made it clearer that this is just a draft with many rough edges. But I first wanted to reach consensus on the general approach before I polish the code.

Oh, well, looked good to me (although I recognized it's a draft). The only other thing I can think of are are tests. E.g., to check the priorities of the different ways of setting configuration options.

**Comment by robinkrahl** [comment]
If we want to restrict the configuration to the top-level options, we could refactor these options into a `GlobalArgs` struct. Then `Args` could use this struct with `#[structopt(flatten)]`, the configuration could be deserialized into this struct too and we could automatically merge the configuration into the CLI arguments using the derived `Merge` trait.

This would allow us to remove the manual merge code.  The drawback is that we would have to make bigger changes if we want to add configuration for subcommand options in the future.  What do you think?

**Comment by robinkrahl** [comment]
Fixed some of the issues, rebased on devel.

File: src/args.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (robinkrahl) [reply src/args.rs:63]: Fixing the macro will be a bit harder than I remembered because it already defines some derives.  As a first step, I simplified the manual Deserialize implementation.

**Comment by d-e-s-o** [comment]
Thanks. Will take a look on the weekend or so.

File: src/args.rs [hunk 9]
```diff
@@ -57,6 +57,18 @@ impl From<DeviceModel> for nitrokey::Model {
...
+  fn deserialize<D>(deserializer: D) -> Result<DeviceModel, D::Error>
+  where
+    D: serde::Deserializer<'de>,
+  {
+    use serde::de::Error;
+
+    let s = String::deserialize(deserializer)?;
+    s.parse().map_err(D::Error::custom)
```
Comment (d-e-s-o) [inline src/args.rs:71]: Could you make that `DeviceModel::from_str` please? It's way easier to understand what is happening this way, I'd say.

File: src/args.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/args.rs:63]: I pretty much thought so. Yeah, that looks fine to me.

**Comment by d-e-s-o** [comment]
I think it looks great!

**Comment by d-e-s-o** [comment]
Make it ready for review when you feel like it.

**Comment by d-e-s-o** [comment]
Btw., feel free to merge commits adding a new crate as a dependency into the first one using it. I had separate commits for that in the past because of the subrepo magic we had (that checked in the respective source code). But without that I'd say it just adds noise and obfuscates why a dependency is introduced. But if you see value then keep it.

File: src/config.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/config.rs:74]: > If we want to restrict the configuration to the top-level options, we could refactor these options into a `GlobalArgs` struct. Then `Args` could use this struct with `#[structopt(flatten)]`, the configuration could be deserialized into this struct too and we could automatically merge the configuration into the CLI arguments using the derived `Merge` trait.
> 
> This would allow us to remove the manual merge code. The drawback is that we would have to make bigger changes if we want to add configuration for subcommand options in the future. What do you think?

I think what you have is fine. We could do that in the future if we really see a need.

[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/config.rs:74]: Ups, that was meant to be a top level comment...

**Comment by robinkrahl** [comment]
Also, I have been thinking about dropping the `directories`/`app-dirs2` dependency and using `xdg` directly:
- On Linux systems, we want to use the XDG Base Directory Specification anyway.
- There is a debate whether command-line applications should use the library paths or the XDG Base Directory Specification on macOS, see e. g. [this discussion](https://stackoverflow.com/questions/3373948/equivalents-of-xdg-config-home-and-xdg-data-home-on-mac-os-x) on stackoverflow.com.
- As far as I know, there are no common conventions for command-line applications on Windows.

Users who prefer the macOS library or Windows AppData paths could still set the XDG environment variables.  What do you think?

**Comment by d-e-s-o** [comment]
> Also, I have been thinking about dropping the `directories`/`app-dirs2` dependency and using `xdg` directly:
> [...]
> Users who prefer the macOS library or Windows AppData paths could still set the XDG environment variables. What do you think?

I'd be fine with that, as long as it still works on those operating systems (which I can't test). So if we were to use `xdg` on, say, Windows (and assuming no configuration on the user's side), the configuration would be expected at `%USERPROFILE%/.config/nitrocli`, is that correct?

**Comment by robinkrahl** [comment]
> I'd be fine with that, as long as it still works on those operating
> systems (which I can't test). So if we were to use `xdg` on, say,
> Windows (and assuming no configuration on the user's side), the
> configuration would be expected at `%USERPROFILE%/.config/nitrocli`,
> is that correct?

I was under the impression that this would be the case as the
documentation of the xdg crate does not mention any platform
restrictions and the lookup algorithm described in the
xdg::BaseDirectories documentation uses std::env::home_dir(), which
falls back to %USERPROFILE% on Windows.

But after looking at the implementation, it turns out that they set
    #![cfg(any(unix, target_os = "redox"))]
in their lib.rs.  So let’s stick with app-dirs2 or directories.

**Comment by robinkrahl** [comment]
Rebased onto topic/anyhow.

**Comment by robinkrahl** [comment]
* Rebased onto devel.
* Use merge 0.1.0 instead of Git version.

I think I addressed all comments, so this should be ready for review and merge.

**Comment by d-e-s-o** [comment]
Awesome! Thanks Robin. I'll have a final look later today and then we should be good to merge.

**Comment by d-e-s-o** [comment]
All merged now. Thanks again!
